# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ãƒ»ã‚¢ãƒ¼ãƒ†ã‚£ã‚¹ãƒˆï¼ˆKey Config Artistï¼‰
extends Node


# å€¤ã¯ãƒœã‚¿ãƒ³ç•ªå·ã€‚ãƒ¬ãƒãƒ¼ã¯ +1000
var key_config = {
	# ä»®æƒ³ã‚­ãƒ¼ï¼ˆï¼‘ï¼‰æ±ºå®šãƒœã‚¿ãƒ³ã€ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸é€ã‚Šãƒœã‚¿ãƒ³
	&"VK_Ok" : -1,
	# ä»®æƒ³ã‚­ãƒ¼ï¼ˆï¼’ï¼‰ã‚­ãƒ£ãƒ³ã‚»ãƒ«ãƒœã‚¿ãƒ³ã€ãƒ¡ãƒ‹ãƒ¥ãƒ¼ãƒœã‚¿ãƒ³
	&"VK_Cancel" : -1,
	# ä»®æƒ³ã‚­ãƒ¼ï¼ˆï¼“ï¼‰ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸æ—©é€ã‚Šãƒœã‚¿ãƒ³
	&"VK_FastForward" : -1,
}

var re_button = RegEx.new()
var re_lever = RegEx.new()

# æ“ä½œã—ãŸãƒœã‚¿ãƒ³ã€€ï¼ˆå¤‰æ•°ã‚’å¢—ã‚„ã—ãŸããªã„ã®ã§ãƒ¬ãƒãƒ¼ã¯ï¼‹ï¼‘ï¼ï¼ï¼ã—ã¦å…¥ã‚Œã‚‹ï¼‰
var button_number = -1
var button_presentation_name = &""

# èµ·å‹•ç›´å¾Œã«ã€€ãƒ¬ãƒãƒ¼ãŒå…¥ã£ãŸçŠ¶æ…‹ã§å§‹ã¾ã‚‹ã“ã¨ãŒã‚ã‚‹ã‹ã‚‰ã€ï¼‘ç§’ãã‚‰ã„ç„¡è¦–ã™ã‚‹ãŸã‚ã®ã‚«ã‚¦ãƒ³ã‚¿ãƒ¼
var counter_of_wait = 0.0
var current_step = 0
# WaitForPrompt, Prompt, WaitForInput, Input, InputOk ã®ï¼•ã¤ã€‚ Wait ã‚’å…¥ã‚Œãªã„ã¨åå¿œéæ•ã«ãªã£ã¦ã—ã¾ã†
var turn_state = &"WaitForPrompt"


# ãƒ†ãƒ­ãƒƒãƒ—ãƒ»ã‚³ãƒ¼ãƒ‡ã‚£ãƒãƒ¼ã‚¿ãƒ¼å–å¾—
func get_telop_coordinator():
	return $"../TelopCoordinator"


# éŸ³æ¥½å®¶å–å¾—
func get_musician():
	return $"../Musician"


# ãƒœã‚¿ãƒ³ãŒé‡è¤‡ã™ã‚‹ã‹ï¼Ÿ
func is_key_duplicated(button_number):
	return button_number in self.key_config.values()


# Called when the node enters the scene tree for the first time.
func _ready():

	# ã“ã®æ–‡å­—åˆ—ãŒã©ã†å¤‰åŒ–ã™ã‚‹ã®ã‹ã€ã•ã£ã±ã‚Šåˆ†ã‹ã‚‰ã‚“
	re_button.compile("Joypad Button (\\d)")
	re_lever.compile("Joypad Motion on Axis (\\d)")


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(_delta):
	pass


# ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°ç”»é¢ã«å…¥ã‚‹
func entry():
	self.get_telop_coordinator().get_node("TextBlock").text = """\
	ï¼Šã€€ï¼Šã€€ï¼Š
	"""


func set_key_ok():
	self.get_telop_coordinator().get_node("TextBlock").text = "ï¼Šã€€ï¼Šã€€ï¼Š"


# ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°ã€€ãƒœã‚¿ãƒ³è¨­å®šãŒæ‹’å¦
func set_key_denied():
	self.get_musician().get_node("SE/ğŸ””ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°ä¸å¯éŸ³").play()
	self.get_telop_coordinator().get_node("TextBlock").text = "ä»–ã®ã‚­ãƒ¼ã‚’é¸ã‚“ã§ãã ã•ã„"


func on_process(_delta):
	pass
